### PR TITLE
[SPARK-19602][SQL][TESTS] Add tests for qualified column names

### PR DESCRIPTION
**Pull request by skambha**
## What changes were proposed in this pull request?
- Add tests covering different scenarios with qualified column names
- Please see Section 2 in the design doc for the various test scenarios [here](https://issues.apache.org/jira/secure/attachment/12854681/Design_ColResolution_JIRA19602.pdf)
- As part of SPARK-19602, changes are made to support three part column name. In order to aid in the review and to reduce the diff, the test scenarios are separated out into this PR.

## How was this patch tested?
- This is a **test only** change. The individual test suites were run successfully.

### CODE REVIEW

**Comment by gatorsmile** [comment]
ok to test

**Comment by gatorsmile** [comment]
Some test cases are negative; some test cases are expected to be supported. Could you split them from the positive test cases? Now some test cases are pretty large.

**Comment by gatorsmile** [comment]
Uh. Forgot one more point. 

Could you move these test cases to our new end-to-end SQL query suite into `SQLQueryTestSuite` ?

**Comment by gatorsmile** [comment]
Thank you for working on this! I like the coverage of qualified column names.

**Comment by gatorsmile** [comment]
retest this please

**Comment by gatorsmile** [comment]
We can move the remaining test cases to `SQLQueryTestSuite `, right?

**Comment by skambha** [comment]
Thanks much Xiao for the review and comments. 

I have made the following changes: 
- Separated out the -ve cases from the +ve cases. 
- Moved positive tests and also the cases that should be supported into the SQLQueryTestSuite framework.  A new test file columnresolution.sql and the corresponding master out file is added. 
- Clean up the ColumnResolutionSuite to remove cases that are covered in the SQLQueryTestSuite
- I have kept the -ve cases in the ColumnResolutionSuite because the exprId shows up in the exception.
- I also wanted to cover a case against a hive serde table so I have kept those tests in the ColumnResolutionSuite

Please advise if we should move any others.  Thanks.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/SQLViewSuite.scala [hunk 1]
```diff
@@ -52,6 +52,19 @@ abstract class SQLViewSuite extends QueryTest with SQLTestUtils {
...
+  test("column resolution scenarios with local temp view") {
+    val df = Seq(2).toDF("i1")
+    df.createOrReplaceTempView("table1")
+    withTempView("table1") {
+      checkAnswer(spark.sql("SELECT table1.* FROM table1"), Row(2))
+      checkAnswer(spark.sql("SELECT * FROM table1"), Row(2))
+      checkAnswer(spark.sql("SELECT i1 FROM table1"), Row(2))
+      checkAnswer(spark.sql("SELECT table1.i1 FROM table1"), Row(2))
+      checkAnswer(spark.sql("SELECT a.i1 FROM table1 AS a"), Row(2))
+      checkAnswer(spark.sql("SELECT i1 FROM table1 AS a"), Row(2))
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/SQLViewSuite.scala]: How about these test cases for temporary views? 

```
-- Test data.
CREATE OR REPLACE TEMPORARY VIEW testData AS SELECT * FROM VALUES
(1, 1), (1, 2), (2, 1), (2, 2), (3, 1), (3, 2), (null, 1), (3, null), (null, null)
AS testData(a, b);
```

Comment (skambha) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/SQLViewSuite.scala]: Sure, let me look at converting these too. Thanks.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/ColumnResolutionSuite.scala [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+ */
+
+package org.apache.spark.sql.hive.execution
+
+import org.apache.spark.sql.{AnalysisException, QueryTest, Row}
+import org.apache.spark.sql.hive.test.TestHiveSingleton
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.test.SQLTestUtils
+
+class ColumnResolutionSuite extends QueryTest with SQLTestUtils with TestHiveSingleton {
```
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/ColumnResolutionSuite.scala]: For the test cases you want to keep here, you can move it to `sql/core`. Why we need to test hive serde tables? Compared with data source tables, it is touching different code paths to resolve columns?

File: sql/core/src/test/resources/sql-tests/inputs/columnresolution.sql [hunk 3]
```diff
@@ -0,0 +1,82 @@
+-- Scenario: column resolution scenarios with datasource table
+CREATE DATABASE mydb1;
+use mydb1;
```
Comment (gatorsmile) [inline sql/core/src/test/resources/sql-tests/inputs/columnresolution.sql]: Please use upper case for SQL keywords.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/SQLViewSuite.scala (hunk 1 above)
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/SQLViewSuite.scala]: Also doable for global temporary view, I think

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/ColumnResolutionSuite.scala (hunk 2 above)
Comment (skambha) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/ColumnResolutionSuite.scala]: The logic to resolve the column in the LogicalPlan is same - there is no change there.  I wanted to test the hive table to make sure that the qualifier information is correctly set. We update the qualifier info in MetastoreRelation so wanted to have coverage for hive table.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/SQLViewSuite.scala (hunk 1 above)
Comment (skambha) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/SQLViewSuite.scala]: Hi Xiao,  I have moved my new local temp view tests and the global temp view tests to the SQLQueryTestSuite framework as well.  Please take a look. Thanks.

**Comment by skambha** [comment]
- Changes to the SQLQueryTestSuite framework to mask the exprId so I can add the -ve cases as well using this framework.
- Added -ve test cases to the SQLQueryTestSuite framework and so removed the hive specific test suite.  For the hive table testcase, I will add that test as part of the actual code changes PR.
- I synced up the codeline and there was one test output inner-join.sql.out that needed a comment to be updated, so I have updated that as well.

File: sql/core/src/test/resources/sql-tests/inputs/columnresolution-negative.sql [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+SELECT mydb1.t1 FROM t1;
+SELECT t1.x.y.* FROM t1;
+SELECT t1 FROM mydb1.t1;
+USE mydb2;
+SELECT mydb1.t1.i1 FROM t1;
+
+-- reset
+set spark.sql.crossJoin.enabled = false;
+DROP DATABASE mydb1 CASCADE;
+DROP DATABASE mydb2 CASCADE;
```
Comment (gatorsmile) [inline sql/core/src/test/resources/sql-tests/inputs/columnresolution-negative.sql]: Please add one empty space after this line

File: sql/core/src/test/resources/sql-tests/inputs/columnresolution-negative.sql [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+-- Negative tests
+USE mydb1;
+SELECT mydb1.t1 FROM t1;
+SELECT t1.x.y.* FROM t1;
+SELECT t1 FROM mydb1.t1;
+USE mydb2;
+SELECT mydb1.t1.i1 FROM t1;
+
+-- reset
+set spark.sql.crossJoin.enabled = false;
```
Comment (gatorsmile) [inline sql/core/src/test/resources/sql-tests/inputs/columnresolution-negative.sql]: Move this to line 24

File: sql/core/src/test/resources/sql-tests/inputs/columnresolution-views.sql [hunk 6]
```diff
@@ -0,0 +1,25 @@
+-- Tests for qualified column names for the view code-path
+-- Test scenario with Temporary view
+CREATE OR REPLACE TEMPORARY VIEW table1 AS SELECT 2 AS i1;
```
Comment (gatorsmile) [inline sql/core/src/test/resources/sql-tests/inputs/columnresolution-views.sql]: `table1 ` -> `view1`

File: sql/core/src/test/resources/sql-tests/inputs/columnresolution-views.sql [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+SELECT * FROM table1;
+SELECT table1.i1 FROM table1;
+SELECT i1 FROM table1;
+SELECT a.i1 FROM table1 AS a;
+SELECT i1 FROM table1 AS a;
+-- cleanup
+DROP VIEW table1;
+
+-- Test scenario with Global Temp view
+CREATE OR REPLACE GLOBAL TEMPORARY VIEW t1 as SELECT 1 as i1;
```
Comment (gatorsmile) [inline sql/core/src/test/resources/sql-tests/inputs/columnresolution-views.sql]: `t1` -> `view1`

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala [hunk 8]
```diff
@@ -228,12 +228,13 @@ class SQLQueryTestSuite extends QueryTest with SharedSQLContext {
...
     } catch {
-      case a: AnalysisException if a.plan.nonEmpty =>
+      case a: AnalysisException =>
         // Do not output the logical plan tree which contains expression IDs.
         // Also implement a crude way of masking expression IDs in the error message
         // with a generic pattern "###".
+        val msg = if (a.plan.nonEmpty) a.getSimpleMessage else a.getMessage
         (StructType(Seq.empty),
-          Seq(a.getClass.getName, a.getSimpleMessage.replaceAll("#\\d+", "#x")))
+          Seq(a.getClass.getName, msg.replaceAll("#\\d+", "#x")))
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala]: Nit: `(StructType(Seq.empty), Seq(a.getClass.getName, msg.replaceAll("#\\d+", "#x")))`

File: sql/core/src/test/resources/sql-tests/results/columnresolution-negative.sql.out [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+Reference 'i1' is ambiguous, could be: i1#x, i1#x.; line 1 pos 7
+
+
+-- !query 9
+SELECT t1.i1 FROM t1, mydb1.t1
+-- !query 9 schema
+struct<>
+-- !query 9 output
+org.apache.spark.sql.AnalysisException
+Reference 't1.i1' is ambiguous, could be: i1#x, i1#x.; line 1 pos 7
```
Comment (gatorsmile) [inline sql/core/src/test/resources/sql-tests/results/columnresolution-negative.sql.out]: To the other reviewer, this is the reason why we need to make a change in `SQLQueryTestSuite.scala`

File: sql/core/src/test/resources/sql-tests/inputs/columnresolution.sql [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+-- TODO: Support this scenario
+SELECT mydb1.t1.i1 FROM t1, mydb1.t1;
+
+-- Scenario: Table with struct column
+USE mydb1;
+CREATE TABLE t5(i1 INT, t5 STRUCT<i1:INT, i2:INT>) USING parquet;
+INSERT INTO t5 VALUES(1, (2, 3));
+SELECT t5.i1 FROM t5;
+SELECT t5.t5.i1 FROM t5;
+SELECT t5.t5.i1 FROM mydb1.t5;
```
Comment (gatorsmile) [inline sql/core/src/test/resources/sql-tests/inputs/columnresolution.sql]: Add two cases for verifying `*`

**Comment by gatorsmile** [comment]
Generally, it looks good to me.

**Comment by gatorsmile** [comment]
LGTM

**Comment by gatorsmile** [comment]
Thanks! Merging to master.

**Comment by skambha** [comment]
Thanks a lot Xiao.

**Pull request closed without merging**